Update ComponentModel\Container stub for component-model 3.1.0 and 4.0.0

jtojnar opened the pull request:
component-model 3.1.0 changes the return type to array when `$deep` argument is `false` (default):
https://github.com/nette/component-model/commit/7f613eed7f5e57b6bde2d0be1bfdbb7e161620b3

It also deprecates the arguments but we cannot add deprecated annotation to those.
https://github.com/nette/component-model/commit/4e0946a788b4ac42ea903b761c693ec7dd083a69

Furthermore, version 4.0.0 changes the return type of `Container::getComponents` to a properly typed array and removes support for extra arguments:
https://github.com/nette/component-model/commit/c40aa8ed45324438100a4d931ed1fb8dc16b4f34

Let’s drop the stub on that version.

jtojnar committed:
Update ComponentModel\Container stub for component-model 3.1.0

That version changes the return type to array when `$deep` argument is `false` (default):
https://github.com/nette/component-model/commit/7f613eed7f5e57b6bde2d0be1bfdbb7e161620b3

It also deprecates the arguments but we cannot add deprecated annotation to those.
https://github.com/nette/component-model/commit/4e0946a788b4ac42ea903b761c693ec7dd083a69

jtojnar committed:
Update ComponentModel\Container stub for component-model 3.1.0

That version changes the return type to array when `$deep` argument is `false` (default):
https://github.com/nette/component-model/commit/7f613eed7f5e57b6bde2d0be1bfdbb7e161620b3

It also deprecates the arguments but we cannot add deprecated annotation to those.
https://github.com/nette/component-model/commit/4e0946a788b4ac42ea903b761c693ec7dd083a69

jtojnar commented:
> It also deprecates the arguments but we cannot add deprecated annotation to those.

There is no `@param-deprecated` annotation AFAICT and the upcoming `#[Deprecated]` attribute RFC [leaves it up for future](https://wiki.php.net/rfc/deprecated_attribute#future_scope).

jtojnar commented:
Looks like this is recognized properly, as I no longer get

```php
  173    Parameter ＃1 $array of function array_filter expects array, Iterator<int|string, Nette\ComponentModel\IComponent> given.  
```

But for some reason it now thinks it returns the union of both branches:

```php
  173    Parameter ＃1 $array of function array_filter expects array, array<int|string, Nette\ComponentModel\IComponent>|Iterator<int|string, Nette\ComponentModel\IComponent> given.  
```

(This is with PHPStan 1.11.1.)

jtojnar commented:
I tried to reproduce it in Playground but it does not fail there: https://phpstan.org/r/50c723fc-6a84-4ca3-945d-fdc4fc5f9fbc

Possibly this is some difference in how stubs are handled?

jtojnar commented:
Okay, added tests, which reproduce the issue when I force the `nette/component-model` to 3.1.0 – for some reason the returned type is union of the both branches of the ternary expression in the `@return`.

```php
1) PHPStan\Type\Nette\ComponentModelContainerDynamicReturnTypeExtensionTest::testFileAsserts with data set "/home/jtojnar/Projects/phpstan-nette/tests/Type/Nette/data/componentModelContainer31.php:16" ('type', '/home/jtojnar/Projects/phpsta...31.php', 'array<int|string, Nette\Compo...onent>', 'array<int|string, Nette\Compo...onent>', 16)
Expected type array<int|string, Nette\ComponentModel\IComponent>, got type array<int|string, Nette\ComponentModel\IComponent>|Iterator<int|string, Nette\ComponentModel\IComponent> in /home/jtojnar/Projects/phpstan-nette/tests/Type/Nette/data/componentModelContainer31.php on line 16.
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'array<int|string, Nette\ComponentModel\IComponent>'
+'array<int|string, Nette\ComponentModel\IComponent>|Iterator<int|string, Nette\ComponentModel\IComponent>'

phar:///home/jtojnar/Projects/phpstan-nette/vendor/phpstan/phpstan/phpstan.phar/src/Testing/TypeInferenceTestCase.php:81
/home/jtojnar/Projects/phpstan-nette/tests/Type/Nette/ComponentModelContainerDynamicReturnTypeExtensionTest.php:35

2) PHPStan\Type\Nette\ComponentModelContainerDynamicReturnTypeExtensionTest::testFileAsserts with data set "/home/jtojnar/Projects/phpstan-nette/tests/Type/Nette/data/componentModelContainer31.php:17" ('type', '/home/jtojnar/Projects/phpsta...31.php', 'array<int|string, Nette\Forms...ainer>', 'array<int|string, Nette\Compo...onent>', 17)
Expected type array<int|string, Nette\Forms\Container>, got type array<int|string, Nette\ComponentModel\IComponent>|Iterator<int|string, Nette\ComponentModel\IComponent> in /home/jtojnar/Projects/phpstan-nette/tests/Type/Nette/data/componentModelContainer31.php on line 17.
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'array<int|string, Nette\Forms\Container>'
+'array<int|string, Nette\ComponentModel\IComponent>|Iterator<int|string, Nette\ComponentModel\IComponent>'

phar:///home/jtojnar/Projects/phpstan-nette/vendor/phpstan/phpstan/phpstan.phar/src/Testing/TypeInferenceTestCase.php:81
/home/jtojnar/Projects/phpstan-nette/tests/Type/Nette/ComponentModelContainerDynamicReturnTypeExtensionTest.php:35

3) PHPStan\Type\Nette\ComponentModelContainerDynamicReturnTypeExtensionTest::testFileAsserts with data set "/home/jtojnar/Projects/phpstan-nette/tests/Type/Nette/data/componentModelContainer31.php:18" ('type', '/home/jtojnar/Projects/phpsta...31.php', 'Iterator<int|string, Nette\Co...onent>', 'array<int|string, Nette\Compo...onent>', 18)
Expected type Iterator<int|string, Nette\ComponentModel\IComponent>, got type array<int|string, Nette\ComponentModel\IComponent>|Iterator<int|string, Nette\ComponentModel\IComponent> in /home/jtojnar/Projects/phpstan-nette/tests/Type/Nette/data/componentModelContainer31.php on line 18.
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'Iterator<int|string, Nette\ComponentModel\IComponent>'
+'array<int|string, Nette\ComponentModel\IComponent>|Iterator<int|string, Nette\ComponentModel\IComponent>'

phar:///home/jtojnar/Projects/phpstan-nette/vendor/phpstan/phpstan/phpstan.phar/src/Testing/TypeInferenceTestCase.php:81
/home/jtojnar/Projects/phpstan-nette/tests/Type/Nette/ComponentModelContainerDynamicReturnTypeExtensionTest.php:35

4) PHPStan\Type\Nette\ComponentModelContainerDynamicReturnTypeExtensionTest::testFileAsserts with data set "/home/jtojnar/Projects/phpstan-nette/tests/Type/Nette/data/componentModelContainer31.php:19" ('type', '/home/jtojnar/Projects/phpsta...31.php', 'Iterator<int|string, Nette\Fo...ainer>', 'array<int|string, Nette\Compo...onent>', 19)
Expected type Iterator<int|string, Nette\Forms\Container>, got type array<int|string, Nette\ComponentModel\IComponent>|Iterator<int|string, Nette\ComponentModel\IComponent> in /home/jtojnar/Projects/phpstan-nette/tests/Type/Nette/data/componentModelContainer31.php on line 19.
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'Iterator<int|string, Nette\Forms\Container>'
+'array<int|string, Nette\ComponentModel\IComponent>|Iterator<int|string, Nette\ComponentModel\IComponent>'

phar:///home/jtojnar/Projects/phpstan-nette/vendor/phpstan/phpstan/phpstan.phar/src/Testing/TypeInferenceTestCase.php:81
/home/jtojnar/Projects/phpstan-nette/tests/Type/Nette/ComponentModelContainerDynamicReturnTypeExtensionTest.php:35

5) PHPStan\Type\Nette\ComponentModelContainerDynamicReturnTypeExtensionTest::testFileAsserts with data set "/home/jtojnar/Projects/phpstan-nette/tests/Type/Nette/data/componentModelContainer31.php:20" ('type', '/home/jtojnar/Projects/phpsta...31.php', 'array<int|string, Nette\Compo...onent>', 'mixed', 20)
Expected type array<int|string, Nette\ComponentModel\IComponent>|Iterator<int|string, Nette\ComponentModel\IComponent>, got type mixed in /home/jtojnar/Projects/phpstan-nette/tests/Type/Nette/data/componentModelContainer31.php on line 20.
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'array<int|string, Nette\ComponentModel\IComponent>|Iterator<int|string, Nette\ComponentModel\IComponent>'
+'mixed'

phar:///home/jtojnar/Projects/phpstan-nette/vendor/phpstan/phpstan/phpstan.phar/src/Testing/TypeInferenceTestCase.php:81
/home/jtojnar/Projects/phpstan-nette/tests/Type/Nette/ComponentModelContainerDynamicReturnTypeExtensionTest.php:35
```

And it seems to work just fine in the playground:

https://phpstan.org/r/d65ff1c0-3cf4-4cda-ab15-792478c7ce34

jtojnar commented:
Also note that the CI currently does not test component-model 3.1.0 due to package and PHP version conflicts. I had to force it locally:

```

--- a/composer.json
+++ b/composer.json
@@ -6,7 +6,8 @@
 		"MIT"
 	],
 	"require": {
-		"php": "^7.4 || ^8.0",
+		"php": " ^8.0",
+		"nette/component-model": "3.1.0",
 		"phpstan/phpstan": "^2.0"
 	},
 	"conflict": {
@@ -21,7 +22,7 @@
 		"nette/application": "^3.0",
 		"nette/di": "^3.1.10",
 		"nette/forms": "^3.0",
-		"nette/utils": "^2.3.0 || ^3.0.0",
+		"nette/utils": "^2.3.0 || ^3.0.0 || ^4.0.0",
 		"php-parallel-lint/php-parallel-lint": "^1.2",
 		"phpstan/phpstan-phpunit": "^2.0",
 		"phpstan/phpstan-strict-rules": "^2.0",
@@ -29,7 +30,6 @@
 	},
 	"config": {
 		"platform": {
-			"php": "7.4.6"
 		},
 		"sort-packages": true
 	},
```

staabm commented on src/Stubs/Nette/StubFilesExtensionLoader.php:
```diff
@@ -0,0 +1,47 @@
+
+	public function getFiles(): array
+	{
+		$stubsDir = dirname(dirname(dirname(__DIR__))) . '/stubs';
+		$path = $stubsDir;
+
+		$files = [];
+
+		$componentModelVersion = self::getInstalledVersion('nette/component-model');
+		if ($componentModelVersion !== null && version_compare($componentModelVersion, '3.1.0', '>=')) {
+			$files[] = $path . '/ComponentModel/Container_3_1.stub';
+		} else {
+			$files[] = $path . '/ComponentModel/Container.stub';
+		}
```
I wonder whether we should load the 2 stubs only on the affected "old nette versions" and don't load any stub if the nette version is recent enough (and ships itself with good enough phpdoc/native-types)

VincentLanglet replied:
It sould indeed copy the logic from the symfony extension https://github.com/phpstan/phpstan-symfony/blob/2.0.x/src/Stubs/Symfony/StubFilesExtensionLoader.php

jtojnar replied:
Yes, that was the plan. Done it now. Though I am still plagued by https://github.com/phpstan/phpstan-nette/pull/142#issuecomment-2589362933

jtojnar committed:
Update ComponentModel\Container stub for component-model 4.0.0

That version changes the return type of `Container::getComponents` to properly typed array and removes support for extra arguments:
https://github.com/nette/component-model/commit/c40aa8ed45324438100a4d931ed1fb8dc16b4f34

Let’s drop the stub on that version.
